cmd/kwil-admin/cmds/snapshot/create.go

```diff
@@ -0,0 +1,145 @@
+// This function downloads the snapshot from Postgres DB, sanitizes it and compresses it.
+func pgDump(ctx context.Context, dbName string, dbUser string, dbHost string, dbPort string, snapshotDir string) error {
+
+	// Check if the snapshot directory exists, if not create it
+	err := os.MkdirAll(snapshotDir, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("failed to create snapshot directory: %w", err)
+	}
+
+	dumpFile := filepath.Join(snapshotDir, "kwildb-snapshot.sql.gz")
+	outputFile, err := os.Create(dumpFile)
+	if err != nil {
+		return fmt.Errorf("failed to create dump file: %w", err)
+	}
+	defer outputFile.Close()
+
+	pgDumpCmd := exec.CommandContext(ctx,
+		"pg_dump",
+		"--dbname", dbName,
+		"--format", "plain",
+		"--schema", "kwild_voting",
+		// "--schema", "kwild_chain", // kwild_chain is not included in this snapshot, as we start from genesis
+		"--schema", "kwild_accts",
+		"--schema", "kwild_internal",
+		"--schema", "ds_*",
+		"-T", "kwild_internal.sentry", // Exclude sentry table
+		"--no-unlogged-table-data",
+		"--no-comments",
+		"--create",
+		"--clean", // drops database first before adding commands to create it
+		"--no-publications",
+		"--no-unlogged-table-data",
+		"--no-tablespaces",
+		"--no-table-access-method",
+		"--no-security-labels",
+		"--no-subscriptions",
+		"--large-objects",
+		"-U", dbUser,
+		"-h", dbHost,
+		"-p", dbPort,
+	)
+
+	sedCmd := exec.CommandContext(ctx, "sed", "-e", "/^--/d", "-e", "/^$/d", "-e", "/^SET/d", "-e", "/^SELECT/d")
+
+	// Create a pipe between pg_dump and sed commands
+	reader, writer := io.Pipe()
+	pgDumpCmd.Stdout = writer
+	sedCmd.Stdin = reader
+	pgDumpCmd.Stderr = os.Stderr
+
+	sedStdoutPipe, err := sedCmd.StdoutPipe()
+	if err != nil {
+		return fmt.Errorf("failed to create stdout pipe from sed command: %w", err)
+	}
+
+	if err := sedCmd.Start(); err != nil {
+		return fmt.Errorf("failed to start sed command: %w", err)
+	}
+
+	if err := pgDumpCmd.Start(); err != nil {
+		return fmt.Errorf("failed to start pg_dump command: %w", err)
+	}
+
+	// Close the writer when pg_dump completes to signal EOF to sed
+	go func() {
+		if err := pgDumpCmd.Wait(); err != nil {
+			fmt.Printf("pg_dump command failed: %v\n", err)
+		}
+		writer.Close()
+	}()
+
+	// Compress the sed output
+	gzipWriter := gzip.NewWriter(outputFile)
+	defer gzipWriter.Close()
+
+	hasher := sha256.New()
+	_, err = io.Copy(io.MultiWriter(hasher, gzipWriter), sedStdoutPipe)
+	if err != nil {
+		return fmt.Errorf("failed to write compressed dump file: %w", err)
+	}
+
+	if err := gzipWriter.Close(); err != nil {
+		return fmt.Errorf("failed to close gzip writer: %w", err)
+	}
+
+	if err := sedCmd.Wait(); err != nil {
+		return fmt.Errorf("failed to wait for sed command: %w", err)
+	}
+
+	// Generate snapshot hash and use it as the genesis hash.
+	hash := hasher.Sum(nil)
+	hashStr := hex.EncodeToString(hash)
```

Comment:
Isn't all of this logic already defined in `internal/snapshots`?

It seems like it would be preferable to rely on that, especially because this seems like something that could change version to version

Reply:
Not actually, the compression part is same, but the pg_dump args and the sanitization process is a bit different.

Reply:
I'm actually more confused by this then. After looking deeper, it seems like this does not perform the same cleaning and determinism logic that the internal engine does. Is this because it is not making it a network snapshot?

I feel like it might still be important to have determinism here, even if it is not strictly required. I would imagine users creating a new network would likely want to compare their starting snapshots that they are using for genesis.

Obviously getting a bit speculative, but just found it surprising that this was not the same process

Reply:
The determinism is needed in Statesync because, the snapshot chunks are coming from multiple nodes and we want literally each byte to be same, so that we can coalesce  chunks from different senders and validate them based off the trusted snapshot providers metadata. 

Whereas in Network migrations, there would be one snapshot file, distributed to other nodes along with the genesis file and the genesis hash. So, it doesn't need that sorting ordering that we do in the statesync snapshots. But we sanitize it to remove the commands that are not required in the sql dump.